simplified-cardcreator/src/main/java/org/nypl/simplified/cardcreator/ui/LocationFragment.kt

```diff
@@ -0,0 +1,263 @@
+    super.onViewCreated(view, savedInstanceState)
+
+    navController = Navigation.findNavController(activity!!, R.id.card_creator_nav_host_fragment)
+    isInNewYorkState()
+
+    binding.checkLocationBtn.setOnClickListener {
+      isInNewYorkState()
+    }
+
+    // Go to next screen
+    binding.nextBtn.setOnClickListener {
+      if (isNewYork) {
+        logger.debug("User navigated to the next screen")
+        nextAction = LocationFragmentDirections.actionNext()
+        navController.navigate(nextAction)
+      } else {
+        val data = Intent()
+        activity!!.setResult(Activity.RESULT_CANCELED, data)
+        activity!!.finish()
+      }
+    }
+
+    // Go to previous screen
+    binding.prevBtn.setOnClickListener {
+      activity!!.onBackPressed()
+    }
+  }
+
+  /**
+   * Show loading screen
+   */
+  private fun showLoading(show: Boolean) {
+    logger.debug("Toggling loading screen")
+    if (show) {
+      binding.loading.visibility = View.VISIBLE
+    } else {
+      binding.loading.visibility = View.GONE
+    }
+  }
+
+  /**
+   * Checks to see if the user has granted required location permissions
+   */
+  private fun getLocation(): Location? {
+    var location: Location? = null
+    logger.debug("Checking for location permission")
+    if (ContextCompat.checkSelfPermission(activity!!, Manifest.permission.ACCESS_COARSE_LOCATION)
+      != PackageManager.PERMISSION_GRANTED
+    ) {
+      logger.debug("Requesting location permission")
+      ActivityCompat.requestPermissions(
+        activity!!,
+        arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION),
+        locationRequestCode
+      )
+    } else {
+      logger.debug("Location permission granted")
+      try {
+        logger.debug("Getting current location")
+        val locationManager = activity!!.getSystemService(Context.LOCATION_SERVICE) as LocationManager
+        val isNetworkLocationEnabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER)
+
+        if (isNetworkLocationEnabled) {
+          locationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, minTimeUpdates, minDistanceUpdates, this)
+          location = locationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER)
+        }
+
+      } catch (e: Exception) {
+        logger.debug("Error getting current location")
+        e.printStackTrace()
+      }
+    }
+    return location
+  }
+
+  /**
+   * Checks to see if a give location is in the state of New York
+   */
+  // TODO: This function is doing to much, break it up into smaller pieces
+  private fun isInNewYorkState(): Boolean {
+    logger.debug("Checking to see if user is in New York")
+    showLoading(true)
+    val locationManager = activity!!.getSystemService(Context.LOCATION_SERVICE) as LocationManager
+    binding.nextBtn.isEnabled = false
+    val maxResults = 1
+    val location = getLocation()
+    val geocoder = Geocoder(activity!!, Locale.getDefault())
```

Comment:
I've had trouble in the past with the Geocoder not working on some devices. We'll want to keep a close eye on this.

Reply:
Yea, pretty much the best we can do without using the [Fused Location Provider API](https://developers.google.com/location-context/fused-location-provider)